### PR TITLE
osutil: pass --extrausers option to groupdel

### PR DESCRIPTION
**Pull request by mardy**
Do not shy off from deleting groups using the `--extrausers` option, if
needed.

See https://bugs.launchpad.net/bugs/1840375

Needs someone (@pedronis ?) to decide what to do on systems where `groupdel` does not support the `--extrausers` option: this branch currently silently ignores the error (meaning that `groupdel` has failed but the user is not aware of it), because this leads to the same outcome we had before this branch. But we might want to return an error instead?

### CODE REVIEW

**Comment by mardy** [comment]
Well, actually, the code is returning an error in any case here, so I guess we can simplify the code and always return the error describing the groupdel failure.

File: osutil/user.go [hunk 1]
```diff
@@ -150,12 +151,14 @@ func EnsureUserGroup(name string, id uint32, extraUsers bool) error {
...
-		if !extraUsers {
-			delCmdStr = append(delCmdStr, name)
-			cmd = exec.Command(delCmdStr[0], delCmdStr[1:]...)
-			if output2, err2 := cmd.CombinedOutput(); err2 != nil {
+		delCmdStr = append(delCmdStr, name)
+		cmd = exec.Command(delCmdStr[0], delCmdStr[1:]...)
+		if output2, err2 := cmd.CombinedOutput(); err2 != nil {
+			if extraUsers && bytes.Contains(output2, []byte("unrecognized option")) {
+				// We have hit https://bugs.launchpad.net/bugs/1840375
+				// TODO: decide what to do.
```
Comment (anonymouse64) [inline osutil/user.go:159]: probably IMHO we should just log an error and return nil. The previous code gave no indication anything was wrong, whereas if we at least log the user will see that some part of the group deletion failed.

**Review by anonymouse64 — APPROVED** [review]
lgtm, but I think we could at least log in the case where it fails with the --extrausers flag

File: osutil/user.go [hunk 2]
```diff
@@ -150,12 +151,13 @@ func EnsureUserGroup(name string, id uint32, extraUsers bool) error {
...
-		if !extraUsers {
-			delCmdStr = append(delCmdStr, name)
-			cmd = exec.Command(delCmdStr[0], delCmdStr[1:]...)
-			if output2, err2 := cmd.CombinedOutput(); err2 != nil {
+		delCmdStr = append(delCmdStr, name)
+		cmd = exec.Command(delCmdStr[0], delCmdStr[1:]...)
+		if output2, err2 := cmd.CombinedOutput(); err2 != nil {
+			if extraUsers && bytes.Contains(output2, []byte("unrecognized option")) {
+				// We have hit https://bugs.launchpad.net/bugs/1840375
+				return fmt.Errorf("groupdel does not support '--extrausers' option (after %s)", useraddErrStr)
```
Comment (anonymouse64) [inline osutil/user.go:159]: after reads a bit weird here, not sure the right way to phrase it is, but usually we will use the pattern `while doing x, error foo happened`, but really these two errors are orthogonal to each other IMHO, so maybe the thing to do here is to make them separate errors, one each line:

```
errors encountered ensuring group foo exists:
- groupdel does not support --extrausers option
- useradd failed with foo
```

Comment (mardy) [reply osutil/user.go:159]: I've read about go's support for nested errors (with the `%w` modifier passed to `Errorf`); do you think it makes sense to use it here?

Comment (anonymouse64) [reply osutil/user.go:159]: As we discussed in the meeting, `%w` is only meant to be used in the case where callers want to inspect an error type, if we did something simple like:

```go
return fmt.Errorf("some syscall failed while doing foo: %v",err)
```

then if `err` was say a os.PathError, the standard library says you can check for that specific error with `os.IsNotExists`, but now we created a new error type from `Errorf` which doesn't match the os.PathError, so that check would fail. The solution from the Go team to this was to introduce `%w` which _wraps_ errors so that if we did this:

```
return fmt.Errorf("some syscall failed while doing foo: %w",err)
```

now the error returned by Errorf will be wrapped such that the os.IsNotExists function would return true since the IsNotExists function now can use the `Unwrap()` and `As()` or `Is()` functions from go's errors package (see https://golang.org/pkg/errors/). 

As was pointed out though, since we use an old go version for some distros, we cannot use the errors package directly, and instead we have to do sneaky tricks like use the xerrors package and define our format strings as constants, since old Go does not know about the `%w` verb. See here for an example: https://github.com/snapcore/snapd/blob/master/bootloader/lkenv/lkenv.go#L264-L275

Comment (anonymouse64) [reply osutil/user.go:159]: but also as explained wrapped errors is not what we want here since:
- the callers of this function probably don't care to inspect the different error cases
- the errors are not related to each other

Comment (mardy) [reply osutil/user.go]: Done!

**Review by anonymouse64 — APPROVED** [review]
+1 from me now on the error message, but bonus points if you fix the existing confusing error message too :smile:

File: osutil/user.go [hunk 3]
```diff
@@ -150,12 +151,15 @@ func EnsureUserGroup(name string, id uint32, extraUsers bool) error {
...
+		delCmdStr = append(delCmdStr, name)
+		cmd = exec.Command(delCmdStr[0], delCmdStr[1:]...)
+		if output2, err2 := cmd.CombinedOutput(); err2 != nil {
+			if extraUsers && bytes.Contains(output2, []byte("unrecognized option")) {
+				// We have hit https://bugs.launchpad.net/bugs/1840375
+				return fmt.Errorf(`errors encountered ensuring user %s exists:
+- %s
+- groupdel does not support '--extrausers' option`, name, useraddErrStr)
+			} else {
 				return fmt.Errorf("groupdel failed with: %s (after %s)", OutputErr(output2, err2), useraddErrStr)
```
[the remark this answers is not in the document]
Comment (mardy) [reply osutil/user.go]: Indeed it was :-)

I updated the branch now.

**Review by anonymouse64 — APPROVED** [review]
still LGTM, thanks for fixing that other error message

File: osutil/user.go [hunk 4]
```diff
@@ -150,14 +151,19 @@ func EnsureUserGroup(name string, id uint32, extraUsers bool) error {
...
-		if !extraUsers {
-			delCmdStr = append(delCmdStr, name)
-			cmd = exec.Command(delCmdStr[0], delCmdStr[1:]...)
-			if output2, err2 := cmd.CombinedOutput(); err2 != nil {
-				return fmt.Errorf("groupdel failed with: %s (after %s)", OutputErr(output2, err2), useraddErrStr)
+		delCmdStr = append(delCmdStr, name)
+		cmd = exec.Command(delCmdStr[0], delCmdStr[1:]...)
+		if output2, err2 := cmd.CombinedOutput(); err2 != nil {
+			var groupdelErrStr string
+			if extraUsers && bytes.Contains(output2, []byte("unrecognized option")) {
```
Comment (mvo5) [inline osutil/user.go:158]: This is tricky - if this code runs localized we will not detect the error. We could always run userdel with "LC_ALL=C.UTF-8" but then the error could be confusing to the user. My feeling is we should just remove the if extraUsers special case and always create the "OutputErr". I.e. something like:
```go
diff --git a/osutil/user.go b/osutil/user.go
index 851d268033..6ebf45e714 100644
--- a/osutil/user.go
+++ b/osutil/user.go
@@ -20,7 +20,6 @@
 package osutil
 
 import (
-       "bytes"
        "fmt"
        "os"
        "os/exec"
@@ -154,13 +153,7 @@ func EnsureUserGroup(name string, id uint32, extraUsers bool) error {
                delCmdStr = append(delCmdStr, name)
                cmd = exec.Command(delCmdStr[0], delCmdStr[1:]...)
                if output2, err2 := cmd.CombinedOutput(); err2 != nil {
-                       var groupdelErrStr string
-                       if extraUsers && bytes.Contains(output2, []byte("unrecognized option")) {
-                               // We have hit https://bugs.launchpad.net/bugs/1840375
-                               groupdelErrStr = "groupdel does not support '--extrausers' option"
-                       } else {
-                               groupdelErrStr = OutputErr(output2, err2).Error()
-                       }
+                       groupdelErrStr := OutputErr(output2, err2).Error()
                        return fmt.Errorf(`errors encountered ensuring user %s exists:
 - %s
 - %s`, name, useraddErrStr, groupdelErrStr)
diff --git a/osutil/user_test.go b/osutil/user_test.go
index 3227670025..996f899137 100644
--- a/osutil/user_test.go
+++ b/osutil/user_test.go
@@ -567,7 +567,7 @@ func (s *ensureUserSuite) TestEnsureUserGroupFailedUseraddCoreNoExtra(c *check.C
        err := osutil.EnsureUserGroup("lakatos", 123456, true)
        c.Assert(err, check.ErrorMatches, `errors encountered ensuring user lakatos exists:
 - useradd failed with: some error
-- groupdel does not support '--extrausers' option`)
+- groupdel: unrecognized option '--extrauser'`)
 
        c.Check(mockGroupDel.Calls(), check.DeepEquals, [][]string{
                {"groupdel", "--extrausers", "lakatos"},
```
and it seems the error is still readable enough(?)

**Review by mvo5 — APPROVED** [review]
Looks fine, just one suggestion for simplification

File: osutil/user.go (hunk 4 above)
Comment (mardy) [reply osutil/user.go:158]: Oh, right! I could just search for the `--extrausers` string, which should be printed regardless of the locale, but indeed all this `if` block is a bit too much for just getting a slightly better error message. So I'll happily apply your suggestion :-)

Comment (mardy) [reply osutil/user.go]: Done!
